x/multisig/msg.go

```diff
@@ -44,19 +40,62 @@ func (UpdateContractMsg) Path() string {
 
 // Validate enforces sigs and threshold boundaries
 func (c *UpdateContractMsg) Validate() error {
-	if len(c.Sigs) == 0 {
-		return errors.Wrap(errors.ErrInvalidMsg, "missing sigs")
+	switch n := len(c.Participants); {
+	case n == 0:
+		return errors.Wrap(errors.ErrInvalidMsg, "no participants")
+	case n > maxParticipantsAllowed:
+		return errors.Wrap(errors.ErrInvalidMsg, "too many participants")
 	}
-	if c.ActivationThreshold <= 0 || int(c.ActivationThreshold) > len(c.Sigs) {
-		return errors.Wrap(errors.ErrInvalidMsg, invalidThreshold)
-	}
-	if c.AdminThreshold <= 0 {
-		return errors.Wrap(errors.ErrInvalidMsg, invalidThreshold)
+	return validateWeights(errors.ErrInvalidMsg,
+		c.Participants, c.ActivationThreshold, c.AdminThreshold)
+}
+
+// validateWeights returns an error if given participants and thresholds
+// configuration is not valid. This check is done on model and messages so
+// instead of copying the code it is extracted into this function.
+func validateWeights(
+	baseErr error,
+	ps []*Participant,
+	activationThreshold Weight,
+	adminThreshold Weight,
+) error {
+	if len(ps) == 0 {
+		return errors.Wrap(baseErr, "missing participants")
 	}
-	for _, a := range c.Sigs {
-		if err := weave.Address(a).Validate(); err != nil {
-			return err
+
+	for _, p := range ps {
+		if err := p.Power.Validate(); err != nil {
+			return errors.Wrapf(err, "participant %s", p.Signature)
+		}
+		if err := p.Signature.Validate(); err != nil {
+			return errors.Wrapf(err, "participant %s", p.Signature)
 		}
 	}
+	if err := activationThreshold.Validate(); err != nil {
+		return errors.Wrap(err, "activation threshold")
+	}
+	if err := adminThreshold.Validate(); err != nil {
+		return errors.Wrap(err, "admin threshold")
+	}
+
+	var total Weight
```

Comment:
req: you need to check max power <256 as in the spec. Otherwise this can give you an overflow

Reply:
Weight is uint32 as there is nothing below 32 bit in protobuf :muscle:

Reply:
Done in https://github.com/iov-one/weave/pull/416/commits/21a8610af61bd88e5c958ce7bee05ea7f934dc0a